### PR TITLE
Using Bullet as a Collision Checker

### PR DESCRIPTION
**Pull request by j-petit**
### Description
This tutorial demonstrate how to use Bullet as a collision checker and is part of the [GSoC 2019](https://github.com/ros-planning/moveit/issues/1427). It shows normal collision detection as well as continuous collision detection. I am using a lot of code from the `Visualizing Collisions` tutorial to have the same interactive robot setup with the visualized collisions.

This does not build and pass yet, as the [feature branch](https://github.com/ros-planning/moveit/tree/feature-bullet-trajopt) has to be merged into master first.

### CODE REVIEW

File: doc/adding_collision_detector/adding_collision_checker.rst
Comment (davetcoleman) [inline doc/adding_collision_detector/adding_collision_checker.rst]: ??

File: doc/adding_collision_detector/src/adding_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/adding_collision_detector/src/adding_collision_checker_tutorial.cpp]: ??

File: doc/adding_collision_detector/adding_collision_checker.rst
Comment (j-petit) [reply doc/adding_collision_detector/adding_collision_checker.rst]: This will be filled as soon as https://github.com/ros-planning/moveit/pull/1551 is merged.

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: please use longer function names for clarity e.g. printHelp()

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: a Marker publisher is already included in MoveItVisualTools (see rviz_visual_tools, the parent class), so you should not need this.

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: a DisplayRobotState publisher is already included in MoveItVisualTools, so you should not need this unless you need three of those publishers...

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]:
```suggestion
    InteractiveRobot interactive_robot;
```

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: replace with moveit_visual_tools::deleteAllMarkers()

https://github.com/PickNikRobotics/rviz_visual_tools/blob/master/include/rviz_visual_tools/rviz_visual_tools.h

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: here also

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (davetcoleman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]:
```suggestion
  Eigen::Isometry3d pose{ Eigen::Isometry3d::Identity() };
```
clearer var name

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (mamoll) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]:
```suggestion
  auto planning_scene = std::make_shared<planning_scene::PlanningScene>(robot_model);
```

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (mamoll) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]:
```suggestion
  auto box = std::make_shared<shapes::Shape>(BOX_SIZE, BOX_SIZE, BOX_SIZE);
```

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (j-petit) [reply doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: Thanks for pointing this out, will update it. I thought that using `auto` is discouraged?

Comment (mamoll) [reply doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: Only when the type can't be easily inferred from context, I thought. It's clear in this case.

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (j-petit) [reply doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: if I use the one from MoveItVisualTools, then all topic names get prefixed by `bullet_collision_detection` which means that I can't use the same topic as the interactive robot uses. This means that I can't show both markers / robot states through the same visualization because they need different topics.

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (j-petit) [reply doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: Same as above comment

**Comment by j-petit** [comment]
I addressed the review comments. However, this PR still requires Bullet to be merged via the feature branch before.

**Comment by mamoll** [comment]
Which PRs have to be merged in the moveit repo before this PR can be merged?

**Comment by j-petit** [comment]
The feature branch `feature-bullet-trajopt` into `master`.

**Comment by mamoll** [comment]
Can you create a PR for that? Or do your existing PRs contain the same commits?

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (mlautman) [inline doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: Is there a reason to use raw pointers here?

Comment (mlautman) [reply doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: Either use smart pointer or give a reason why we are using raw pointers here

**Comment by j-petit** [comment]
> Can you create a PR for that? Or do your existing PRs contain the same commits?

We first have to discuss how we deal with the fact that the Bullet version available through `apt` in Xenial is outdated and does not support the collision checking before we can do any merging. See  [this issue](https://github.com/ros-planning/moveit/issues/1570). I added this topic to the agenda for the maintainer meeting next month.

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (j-petit) [reply doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: I adapted the code from [the visualizing tutorial](https://github.com/ros-planning/moveit_tutorials/blob/master/doc/visualizing_collisions/src/visualizing_collisions_tutorial.cpp) for this tutorial and as it uses raw pointers I thought it would be fine.

**Comment by v4hn** [comment]
> We first have to discuss how we deal with the fact that the Bullet version available through apt in Xenial is outdated

In the issue you linked, the proposed solution was to check the bullet versions at compile time and just disable the feature if bullet is too old. Personally, I don't see the problem with that.

**Comment by j-petit** [comment]
So this means enabling/disabling this in the cmake files right? Then I can provide a PR for this.

**Comment by mamoll** [comment]
> So this means enabling/disabling this in the cmake files right? Then I can provide a PR for this.

Yes, typically, you'd have something like this:
```
find_package(bullet 2.88)
set(MOVEIT_HAVE_BULLET bullet_FOUND)
# this file contains: \cmakedefine01 MOVEIT_HAVE_BULLET
configure_file(some_header.h.in some_header.h)
```
If Bullet doesn't include cmake config files or doesn't support version number checking, you have to somehow get the version some other way.

**Comment by j-petit** [comment]
Thanks Mark!

Mark Moll <notifications@github.com> schrieb am Mo., 18. Nov. 2019, 17:15:

**Comment by j-petit** [comment]
Bullet does `set ( BULLET_VERSION_STRING "2.87" )` in the `BulletConfig.cmake`. Do I get this correctly that this is not compatible with the cmake `find_package(Bullet [version])` command? When I do this, it does not matter which number I put in the version. I am not so experienced with setting up a build system with cmake.

**Comment by j-petit** [comment]
All further discussions concerning Bullet and cmake happening [here](https://github.com/ros-planning/moveit/pull/1744).

**Comment by felixvd** [comment]
Would be too bad if this was left by the roadside. It has had a birthday already. Do you think you can give this another push @j-petit ?

**Comment by j-petit** [comment]
wups, completely missed that this is still pending. Thanks for reminding me Felix! I'll try fixing those loose ends in the near future. As I am not working in the ROS environment anymore, it's hard to get started again...

File: doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp
Comment (j-petit) [reply doc/bullet_collision_checker/src/bullet_collision_checker_tutorial.cpp]: I replaced all raw pointers with smart ones.

**Comment by j-petit** [comment]
So, this is finally ready to be merged from my side. ping @felixvd, @mamoll

**Review by mamoll — APPROVED** [review]
This looks good to me. Let's get this merged (finally!).

**Comment by j-petit** [comment]
Fixed Travis, can we get this merged, @mamoll ?

**Comment by mamoll** [comment]
> Fixed Travis, can we get this merged, @mamoll ?

IMO, yes. I'd like to get the approval from anther maintainer. @davetcoleman or @felixvd, perhaps?

**Comment by j-petit** [comment]
Or @mlautman ?

**Comment by felixvd** [comment]
I would like to actually test it before I approve it, but I probably won't get around to it before ROS World :(

**Comment by j-petit** [comment]
That's fine, I think it would be great if someone is actually testing it beforehand!

**Comment by AndyZe** [comment]
As you know, I've started using Bullet recently. It's great so far!

I downloaded the tutorial, built the webpage, and ran it. It all looks good. Let's finally get this merged.

**Comment by j-petit** [comment]
Thanks Andy!

**Pull request merged**
